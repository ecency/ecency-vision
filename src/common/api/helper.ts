import defaults from "../constants/defaults.json";

export const apiBase = (endpoint: string): string => `${defaults.base}${endpoint}`;

export const apiBaseImage = (endpoint: string): string => `${defaults.imageServer}${endpoint}`;
