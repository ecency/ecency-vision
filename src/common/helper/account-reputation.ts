const isHumanReadable = (input: number): boolean => {
  return Math.abs(input) > 0 && Math.abs(input) <= 100;
};

export default (input: string | number): number => {
  if (typeof input === 'number' && isHumanReadable(input)) {
    return Math.floor(input);
  }

  if (typeof input === 'string') {
    input = Number(input);

    if (isHumanReadable(input)) {
      return Math.floor(input);
    }
  }

  if (input === 0) {
    return 25;
  }

  let neg = false;

  if (input < 0) neg = true;

  let reputationLevel = Math.log10(Math.abs(input));
  reputationLevel = Math.max(reputationLevel - 9, 0);

  if (reputationLevel < 0) reputationLevel = 0;

  if (neg) reputationLevel *= -1;

  reputationLevel = reputationLevel * 9 + 25;

  return Math.floor(reputationLevel);
};
