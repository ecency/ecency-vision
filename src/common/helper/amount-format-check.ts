export default (v: string) => /^\d{1,3}(,?\d{3})*(\.(\d{3},?)*\d{1,3})?$/.test(v);
