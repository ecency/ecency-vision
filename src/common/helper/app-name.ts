import stripTags from "../util/strip-tags";

interface ObjInput {
  name?: string;
}

export default (input: string | null | undefined | ObjInput) => {
  if (!input) {
    return "";
  }

  if (typeof input === "string") {
    return stripTags(input);
  }

  if (typeof input === "object" && input.name !== undefined) {
    return stripTags(input.name);
  }

  return "";
};
