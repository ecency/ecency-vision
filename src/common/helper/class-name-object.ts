export function classNameObject(obj: Record<string, boolean | undefined>): string {
  return Object.entries(obj)
    .filter(([className, isActive]) => !!isActive)
    .map(([className]) => className)
    .join(" ");
}
