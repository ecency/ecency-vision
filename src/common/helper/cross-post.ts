import {postBodySummary} from '@ecency/render-helper';
import {Entry} from '../store/entries/types';

export const crossPostMessage = (body: string) => {
  const crossPostRegex =
    /^This is a cross post of \[@(.*?)\/(.*?)\]\(\/.*?@.*?\/.*?\) by @.*?\.<br>/;

  if (body.match(crossPostRegex)) {
    const message = body.replace(crossPostRegex, '');
    return postBodySummary(message);
  }

  return null;
};

export const makeCrossPostMessage = (
  entry: Entry,
  poster: string,
  message: string,
) => {
  return `This is a cross post of [@${entry.author}/${entry.permlink}](/${entry.category}/@${entry.author}/${entry.permlink}) by @${poster}.<br><br>${message}`;
};
