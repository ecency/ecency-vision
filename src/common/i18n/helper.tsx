import React from "react";

import xss from "xss";

import { _t } from "./index";

export const HTML_WHITELIST = {
  a: ["class", "href", "target", "rel"],
  span: ["class"],
  h3: [],
  p: [],
  strong: [],
  b: [],
  img: ["src", "alt"],
  ul: [],
  li: [],
  br: []
};

export const safeHtml = (html: string): string => {
  return xss(html, {
    whiteList: HTML_WHITELIST,
    stripIgnoreTagBody: true,
    stripIgnoreTag: true,
    css: false
  });
};

interface Props {
  k: string;
  args?: {};
  children: JSX.Element;
}

export class Tsx extends React.Component<Props> {
  render() {
    const { k, args } = this.props;

    const { children } = this.props;

    const html = safeHtml(_t(k, args));

    return React.cloneElement(children, { dangerouslySetInnerHTML: { __html: html } });
  }
}
