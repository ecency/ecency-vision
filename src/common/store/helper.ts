import {Store} from 'redux';

import i18n from 'i18next';

import {AppState} from './index';
import {ActiveUser, UserPoints} from './active-user/types';
import {
  loginAct as loginActiveUser,
  logoutAct as logoutActiveUser,
  updateAct as updateActiveUserAct,
} from './active-user';

import {getAccount, getDynamicProps} from '../api/hive';
import {getPoints, usrActivity, getPromotedEntries} from '../api/private-api';
import {reloadAct as reloadUsers} from './users';
import {fetchedAct as loadDynamicProps} from './dynamic-props';
import {fetchedAct as entriesFetchedAct} from './entries';
import {
  setCurrencyAct as setCurrency,
  muteNotificationsAct as muteNotifications,
  setLangAct as setLang,
  setNsfwAct as setNsfw,
} from './global';

import {getCurrencyRate} from '../api/misc';

import currencies from '../constants/currencies.json';

import * as ls from '../../common/util/local-storage';

import currencySymbol from '../../common/helper/currency-symbol';

import {AppWindow} from '../../client/window';

declare var window: AppWindow;

export const activeUserMaker = (
  name: string,
  points: string = '0.000',
  uPoints: string = '0.000',
): ActiveUser => {
  return {
    username: name,
    data: {name},
    points: {
      points,
      uPoints,
    },
  };
};

export const activeUserUpdater = async (store: Store<AppState>) => {
  const state = store.getState();
  if (state.activeUser) {
    const {username} = state.activeUser;

    let account;
    try {
      account = await getAccount(username);
    } catch (e) {
      return;
    }

    let points: UserPoints;
    try {
      const p = await getPoints(username);
      points = {points: p.points, uPoints: p.unclaimed_points};
    } catch (e) {
      points = {
        points: '0.000',
        uPoints: '0.000',
      };
    }

    store.dispatch(updateActiveUserAct(account, points));
  }
};

export const syncActiveUser = (store: Store<AppState>) => {
  const state = store.getState();

  const activeUser = ls.get('active_user');

  // logout
  if (!activeUser && state.activeUser) {
    store.dispatch(logoutActiveUser());
    return;
  }

  // login
  if (activeUser && !state.activeUser) {
    store.dispatch(loginActiveUser());
    activeUserUpdater(store).then();
    return;
  }

  // switch
  if (
    activeUser &&
    state.activeUser &&
    activeUser !== state.activeUser.username
  ) {
    store.dispatch(loginActiveUser());
    activeUserUpdater(store).then();
    return;
  }
};

export const clientStoreTasks = (store: Store<AppState>) => {
  // To use in places where we can't access to store
  window.usePrivate = store.getState().global.usePrivate;

  // Initial state from browser's local storage
  store.dispatch(reloadUsers());
  store.dispatch(loginActiveUser());

  // Load dynamic props
  getDynamicProps().then(resp => {
    store.dispatch(loadDynamicProps(resp));
  });

  // Update active user in interval
  activeUserUpdater(store).then();
  setInterval(() => {
    activeUserUpdater(store).then();
  }, 60 * 1000);

  // Active user sync between tabs / windows
  setInterval(() => {
    syncActiveUser(store);
  }, 2000);

  // Do check-in in interval
  const checkIn = () => {
    const state = store.getState();
    if (state.activeUser) {
      usrActivity(state.activeUser?.username!, 10).then();
    }
  };

  // wait for initial user sync
  setTimeout(() => {
    checkIn();
  }, 5000);
  setInterval(checkIn, 1000 * 60 * 15 + 8);

  // Inject / update promoted entries to store
  const promotedEntries = () => {
    getPromotedEntries().then(r => {
      store.dispatch(entriesFetchedAct('__promoted__', r, false));
    });
  };
  promotedEntries();
  setInterval(promotedEntries, 1000 * 60 * 5);

  // Currency
  const currency = ls.get('currency');
  if (currency && currencies.find(x => x.id === currency)) {
    const symbol = currencySymbol(currency);
    getCurrencyRate(currency).then(rate => {
      store.dispatch(setCurrency(currency, rate, symbol));
    });
  }

  // Notifications
  if (ls.get('notifications') === false) {
    store.dispatch(muteNotifications());
  }

  // Language
  const lang = ls.get('lang');
  if (lang) {
    if (store.getState().global.lang !== lang) {
      i18n.changeLanguage(lang).then(() => {
        store.dispatch(setLang(lang));
      });
    }
  }

  // NSFW
  if (ls.get('nsfw') === true) {
    store.dispatch(setNsfw(true));
  }
};
