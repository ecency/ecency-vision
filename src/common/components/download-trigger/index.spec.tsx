import React from "react";

import renderer from "react-test-renderer";

import { DialogContent } from "./index";

import { allOver } from "../../helper/test-helper";

let MOCK: string = "";

Object.defineProperty(global.navigator, "appVersion", {
  get: () => {
    if (MOCK === "MacOS") {
      return "5.0 (Macintosh; Intel Mac OS X 10_15_2) AppleWebKit/537.36";
    }

    if (MOCK === "WindowsOS") {
      return "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/60.0.3112.113 Safari/537.36";
    }

    if (MOCK === "LinuxOS") {
      return "Mozilla/5.0 (X11; Linux x86_64; rv:52.0) Gecko/20100101 Firefox/52.0";
    }

    if (MOCK === "AndroidOS") {
      return "Mozilla/5.0 (Linux; Android 6.0.1; RedMi Note 5 Build/RB3N5C; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/68.0.3440.91 Mobile Safari/537.36";
    }

    if (MOCK === "iOS") {
      return "Mozilla/5.0 (iPhone; CPU iPhone OS 10_0 like Mac OS X) AppleWebKit/602.1.38 (KHTML, like Gecko) Version/10.0 Mobile/14A5297c Safari/602.1";
    }

    return "";
  }
});

jest.mock("../../api/misc", () => ({
  geLatestDesktopTag: () =>
    new Promise((resolve) => {
      resolve("3.0.16");
    })
}));

it("(1) Mac.", async () => {
  MOCK = "MacOS";
  const component = await renderer.create(<DialogContent />);
  await allOver();
  expect(component.toJSON()).toMatchSnapshot();
});

it("(2) Windows.", async () => {
  MOCK = "WindowsOS";
  const component = await renderer.create(<DialogContent />);
  await allOver();
  expect(component.toJSON()).toMatchSnapshot();
});

it("(3) Linux.", async () => {
  MOCK = "LinuxOS";
  const component = await renderer.create(<DialogContent />);
  await allOver();
  expect(component.toJSON()).toMatchSnapshot();
});

it("(4) Android.", async () => {
  MOCK = "AndroidOS";
  const component = await renderer.create(<DialogContent />);
  await allOver();
  expect(component.toJSON()).toMatchSnapshot();
});

it("(5) iOS.", async () => {
  MOCK = "iOS";
  const component = await renderer.create(<DialogContent />);
  await allOver();
  expect(component.toJSON()).toMatchSnapshot();
});

it("(6) Other.", async () => {
  MOCK = "";
  const component = await renderer.create(<DialogContent />);
  await allOver();
  expect(component.toJSON()).toMatchSnapshot();
});
