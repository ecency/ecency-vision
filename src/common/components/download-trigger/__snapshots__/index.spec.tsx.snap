// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(1) Mac. 1`] = `
<div
  className="download-dialog-content"
>
  <h2
    className="downloads-title"
  >
    Download
  </h2>
  <div
    className="downloads-text"
  >
    Enjoy Weebecash for iPhone, iPad and Android, as well as Windows, Mac or Linux devices
  </div>
  <div
    className="download-buttons"
  >
    <a
      className="download-button btn-desktop"
      href="https://github.com/ecency/ecency-vision/releases/download/3.0.16/Ecency-3.0.16.dmg"
      rel="noopener noreferrer"
      target="_blank"
    >
      Mac
    </a>
    <a
      className="download-button btn-android"
      href="https://android.ecency.com"
      rel="noopener noreferrer"
      target="_blank"
    >
      Android
    </a>
    <a
      className="download-button btn-ios"
      href="https://ios.ecency.com"
      rel="noopener noreferrer"
      target="_blank"
    >
      iOS
    </a>
  </div>
</div>
`;

exports[`(2) Windows. 1`] = `
<div
  className="download-dialog-content"
>
  <h2
    className="downloads-title"
  >
    Download
  </h2>
  <div
    className="downloads-text"
  >
    Enjoy Weebecash for iPhone, iPad and Android, as well as Windows, Mac or Linux devices
  </div>
  <div
    className="download-buttons"
  >
    <a
      className="download-button btn-desktop"
      href="https://github.com/ecency/ecency-vision/releases/download/3.0.16/Ecency-Setup-3.0.16.exe"
      rel="noopener noreferrer"
      target="_blank"
    >
      Windows
    </a>
    <a
      className="download-button btn-android"
      href="https://android.ecency.com"
      rel="noopener noreferrer"
      target="_blank"
    >
      Android
    </a>
    <a
      className="download-button btn-ios"
      href="https://ios.ecency.com"
      rel="noopener noreferrer"
      target="_blank"
    >
      iOS
    </a>
  </div>
</div>
`;

exports[`(3) Linux. 1`] = `
<div
  className="download-dialog-content"
>
  <h2
    className="downloads-title"
  >
    Download
  </h2>
  <div
    className="downloads-text"
  >
    Enjoy Weebecash for iPhone, iPad and Android, as well as Windows, Mac or Linux devices
  </div>
  <div
    className="download-buttons"
  >
    <a
      className="download-button btn-desktop"
      href="https://github.com/ecency/ecency-vision/releases/download/3.0.16/ecency-surfer-3.0.16.tar.gz"
      rel="noopener noreferrer"
      target="_blank"
    >
      Linux
    </a>
    <a
      className="download-button btn-android"
      href="https://android.ecency.com"
      rel="noopener noreferrer"
      target="_blank"
    >
      Android
    </a>
    <a
      className="download-button btn-ios"
      href="https://ios.ecency.com"
      rel="noopener noreferrer"
      target="_blank"
    >
      iOS
    </a>
  </div>
</div>
`;

exports[`(4) Android. 1`] = `
<div
  className="download-dialog-content"
>
  <h2
    className="downloads-title"
  >
    Download
  </h2>
  <div
    className="downloads-text"
  >
    Enjoy Weebecash for iPhone, iPad and Android, as well as Windows, Mac or Linux devices
  </div>
  <div
    className="download-buttons"
  >
    <a
      className="download-button btn-android"
      href="https://android.ecency.com"
      rel="noopener noreferrer"
      target="_blank"
    >
      Android
    </a>
  </div>
</div>
`;

exports[`(5) iOS. 1`] = `
<div
  className="download-dialog-content"
>
  <h2
    className="downloads-title"
  >
    Download
  </h2>
  <div
    className="downloads-text"
  >
    Enjoy Weebecash for iPhone, iPad and Android, as well as Windows, Mac or Linux devices
  </div>
  <div
    className="download-buttons"
  >
    <a
      className="download-button btn-ios"
      href="https://ios.ecency.com"
      rel="noopener noreferrer"
      target="_blank"
    >
      iOS
    </a>
  </div>
</div>
`;

exports[`(6) Other. 1`] = `
<div
  className="download-dialog-content"
>
  <h2
    className="downloads-title"
  >
    Download
  </h2>
  <div
    className="downloads-text"
  >
    Enjoy Weebecash for iPhone, iPad and Android, as well as Windows, Mac or Linux devices
  </div>
  <div
    className="download-buttons"
  >
    <a
      className="download-button btn-desktop"
      href="https://github.com/ecency/ecency-vision/releases/download/3.0.16/Ecency-Setup-3.0.16.exe"
      rel="noopener noreferrer"
      target="_blank"
    >
      Windows
    </a>
    <a
      className="download-button btn-android"
      href="https://android.ecency.com"
      rel="noopener noreferrer"
      target="_blank"
    >
      Android
    </a>
    <a
      className="download-button btn-ios"
      href="https://ios.ecency.com"
      rel="noopener noreferrer"
      target="_blank"
    >
      iOS
    </a>
  </div>
</div>
`;
